TestPeerConnection_Media_Shutdown causes timeout on CI

```
=== RUN   TestPeerConnection_Media_Shutdown
goroutine profile: total 14
1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x8110dfb 0x8113c4a 0x8110aca 0x8111e6d 0x811111b 0x843ef74 0x80754d2 0x809e321
#	0x8110dfa	testing.(*T).Run+0x2ca		/usr/local/go/src/testing/testing.go:961
#	0x8113c49	testing.runTests.func1+0x59	/usr/local/go/src/testing/testing.go:1202
#	0x8110ac9	testing.tRunner+0x99		/usr/local/go/src/testing/testing.go:909
#	0x8111e6c	testing.runTests+0x22c		/usr/local/go/src/testing/testing.go:1200
#	0x811111a	testing.(*M).Run+0x13a		/usr/local/go/src/testing/testing.go:1117
#	0x843ef73	main.main+0x103			_testmain.go:300
#	0x80754d1	runtime.main+0x201		/usr/local/go/src/runtime/proc.go:203

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82b5f62 0x82f1d15 0x82fa4af 0x809e321
#	0x82b5f61	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x82f1d14	github.com/pion/ice.(*Conn).Read+0x74				/go/pkg/mod/github.com/pion/ice@v0.7.7/transport.go:79
#	0x82fa4ae	github.com/pion/webrtc/v2/internal/mux.(*Mux).readLoop+0x9e	/go/src/github.com/pion/webrtc/internal/mux/mux.go:106

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82b5f62 0x82f9d3c 0x8268829 0x809e321
#	0x82b5f61	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x82f9d3b	github.com/pion/webrtc/v2/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:36
#	0x8268828	github.com/pion/dtls/v2.(*Conn).inboundLoop+0x98		/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.5/conn.go:558

1 @ 0x80758a4 0x8070524 0x806f9fb 0x80be807 0x80bf396 0x80bf37b 0x81bb5bf 0x81cf40f 0x81ce1d7 0x82e8ff0 0x809e321
#	0x806f9fa	internal/poll.runtime_pollWait+0x4a			/usr/local/go/src/runtime/netpoll.go:184
#	0x80be806	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x80bf395	internal/poll.(*pollDesc).waitRead+0x165		/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x80bf37a	internal/poll.(*FD).ReadFrom+0x14a			/usr/local/go/src/internal/poll/fd_unix.go:219
#	0x81bb5be	net.(*netFD).readFrom+0x3e				/usr/local/go/src/net/fd_unix.go:208
#	0x81cf40e	net.(*UDPConn).readFrom+0x3e				/usr/local/go/src/net/udpsock_posix.go:47
#	0x81ce1d6	net.(*UDPConn).ReadFrom+0x46				/usr/local/go/src/net/udpsock.go:121
#	0x82e8fef	github.com/pion/ice.(*candidateBase).recvLoop+0xef	/go/pkg/mod/github.com/pion/ice@v0.7.7/candidate_base.go:92

1 @ 0x80758a4 0x8070524 0x806f9fb 0x80be807 0x80bff9d 0x80bff7e 0x81c9a88 0x82c284c 0x82c5e52 0x82c5e31 0x82c9b11 0x809e321
#	0x806f9fa	internal/poll.runtime_pollWait+0x4a			/usr/local/go/src/runtime/netpoll.go:184
#	0x80be806	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x80bff9c	internal/poll.(*pollDesc).waitRead+0xec			/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x80bff7d	internal/poll.(*FD).RawRead+0xcd			/usr/local/go/src/internal/poll/fd_unix.go:534
#	0x81c9a87	net.(*rawConn).Read+0x47				/usr/local/go/src/net/rawconn.go:43
#	0x82c284b	golang.org/x/net/internal/socket.(*Conn).recvMsg+0x18b	/go/pkg/mod/golang.org/x/net@v0.0.0-20191209160850-c0dbc17a3553/internal/socket/rawconn_msg.go:32
#	0x82c5e51	golang.org/x/net/internal/socket.(*Conn).RecvMsg+0x141	/go/pkg/mod/golang.org/x/net@v0.0.0-20191209160850-c0dbc17a3553/internal/socket/socket.go:255
#	0x82c5e30	golang.org/x/net/ipv4.(*payloadHandler).ReadFrom+0x120	/go/pkg/mod/golang.org/x/net@v0.0.0-20191209160850-c0dbc17a3553/ipv4/payload_cmsg.go:31
#	0x82c9b10	github.com/pion/mdns.(*Conn).start+0xf0			/go/pkg/mod/github.com/pion/mdns@v0.0.4/conn.go:256

1 @ 0x80758a4 0x8083e59 0x826616e 0x8266969 0x83ebfed 0x83fcc4d 0x809e321
#	0x826616d	github.com/pion/dtls/v2.createConn+0x7ad				/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.5/conn.go:202
#	0x8266968	github.com/pion/dtls/v2.Server+0x58					/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.5/conn.go:252
#	0x83ebfec	github.com/pion/webrtc/v2.(*DTLSTransport).Start+0x38c			/go/src/github.com/pion/webrtc/dtlstransport.go:282
#	0x83fcc4c	github.com/pion/webrtc/v2.(*PeerConnection).startTransports+0x1ac	/go/src/github.com/pion/webrtc/peerconnection.go:1871

1 @ 0x80758a4 0x8083e59 0x827a768 0x809e321
#	0x827a767	github.com/pion/dtls/v2.(*Conn).startHandshakeOutbound.func1+0xf7	/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.5/conn.go:718

1 @ 0x80758a4 0x8083e59 0x82e551f 0x82e6a05 0x83eea68 0x83efc8b 0x83fa3a2 0x8411ceb 0x8110aca 0x809e321
#	0x82e551e	github.com/pion/ice.(*Agent).run+0x9e					/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:732
#	0x82e6a04	github.com/pion/ice.(*Agent).Close+0x74					/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:929
#	0x83eea67	github.com/pion/webrtc/v2.(*ICEGatherer).Close+0x77			/go/src/github.com/pion/webrtc/icegatherer.go:181
#	0x83efc8a	github.com/pion/webrtc/v2.(*ICETransport).Stop+0x7a			/go/src/github.com/pion/webrtc/icetransport.go:162
#	0x83fa3a1	github.com/pion/webrtc/v2.(*PeerConnection).Close+0x2e1			/go/src/github.com/pion/webrtc/peerconnection.go:1580
#	0x8411cea	github.com/pion/webrtc/v2.TestPeerConnection_Media_Shutdown+0x89a	/go/src/github.com/pion/webrtc/peerconnection_media_test.go:294
#	0x8110ac9	testing.tRunner+0x99							/usr/local/go/src/testing/testing.go:909

1 @ 0x80758a4 0x8083e59 0x82e551f 0x82e92d2 0x82e8fc6 0x809e321
#	0x82e551e	github.com/pion/ice.(*Agent).run+0x9e			/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:732
#	0x82e92d1	github.com/pion/ice.handleInboundCandidateMsg+0x2a1	/go/pkg/mod/github.com/pion/ice@v0.7.7/candidate_base.go:112
#	0x82e8fc5	github.com/pion/ice.(*candidateBase).recvLoop+0xc5	/go/pkg/mod/github.com/pion/ice@v0.7.7/candidate_base.go:97

1 @ 0x80758a4 0x8083e59 0x82e551f 0x82f5efe 0x809e321
#	0x82e551e	github.com/pion/ice.(*Agent).run+0x9e				/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:732
#	0x82f5efd	github.com/pion/ice.(*controllingSelector).Start.func1+0xed	/go/pkg/mod/github.com/pion/ice@v0.7.7/selection.go:36

1 @ 0x80758a4 0x8083e59 0x82e565b 0x809e321
#	0x82e565a	github.com/pion/ice.(*Agent).taskLoop+0xda	/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:743

1 @ 0x80758a4 0x8083e59 0x82f1b4a 0x83efb0e 0x83efabc 0x83fcb21 0x809e321
#	0x82f1b49	github.com/pion/ice.(*Agent).connect+0x129				/go/pkg/mod/github.com/pion/ice@v0.7.7/transport.go:57
#	0x83efb0d	github.com/pion/ice.(*Agent).Dial+0x36d					/go/pkg/mod/github.com/pion/ice@v0.7.7/transport.go:16
#	0x83efabb	github.com/pion/webrtc/v2.(*ICETransport).Start+0x31b			/go/src/github.com/pion/webrtc/icetransport.go:123
#	0x83fcb20	github.com/pion/webrtc/v2.(*PeerConnection).startTransports+0x80	/go/src/github.com/pion/webrtc/peerconnection.go:1856

1 @ 0x80758a4 0x808496d 0x808495e 0x80846ba 0x80a8576 0x83efdc3 0x8425dd4 0x82e5059 0x82e4f77 0x82f095c 0x82e84ba 0x82f4db4 0x82e567f 0x809e321
#	0x80846b9	sync.runtime_SemacquireMutex+0x39						/usr/local/go/src/runtime/sema.go:71
#	0x80a8575	sync.(*RWMutex).RLock+0x55							/usr/local/go/src/sync/rwmutex.go:50
#	0x83efdc2	github.com/pion/webrtc/v2.(*ICETransport).onSelectedCandidatePairChange+0x22	/go/src/github.com/pion/webrtc/icetransport.go:176
#	0x8425dd3	github.com/pion/webrtc/v2.(*ICETransport).Start.func2+0x143			/go/src/github.com/pion/webrtc/icetransport.go:104
#	0x82e5058	github.com/pion/ice.(*Agent).onSelectedCandidatePairChange+0x168		/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:582
#	0x82e4f76	github.com/pion/ice.(*Agent).setSelectedPair+0x86				/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:641
#	0x82f095b	github.com/pion/ice.(*controllingSelector).HandleSuccessResponse+0x3eb		/go/pkg/mod/github.com/pion/ice@v0.7.7/selection.go:176
#	0x82e84b9	github.com/pion/ice.(*Agent).handleInbound+0xc59				/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:1108
#	0x82f4db3	github.com/pion/ice.handleInboundCandidateMsg.func1+0x43			/go/pkg/mod/github.com/pion/ice@v0.7.7/candidate_base.go:113
#	0x82e567e	github.com/pion/ice.(*Agent).taskLoop+0xfe					/go/pkg/mod/github.com/pion/ice@v0.7.7/agent.go:750

1 @ 0x814610c 0x8145f7a 0x8142fd3 0x83e7897 0x809e321
#	0x814610b	runtime/pprof.writeRuntimeProfile+0x7b			/usr/local/go/src/runtime/pprof/pprof.go:708
#	0x8145f79	runtime/pprof.writeGoroutine+0x79			/usr/local/go/src/runtime/pprof/pprof.go:670
#	0x8142fd2	runtime/pprof.(*Profile).WriteTo+0x2c2			/usr/local/go/src/runtime/pprof/pprof.go:329
#	0x83e7896	github.com/pion/transport/test.TimeOut.func1+0x56	/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:18

panic: timeout

goroutine 1319 [running]:
github.com/pion/transport/test.TimeOut.func1()
	/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:21 +0xda
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x37
FAIL	github.com/pion/webrtc/v2	65.093s
```

--- a/icetransport.go
+++ b/icetransport.go
@@ -16,7 +16,8 @@ import (
 // ICETransport allows an application access to information about the ICE
 // transport over which packets are sent and received.
 type ICETransport struct {
-	lock sync.RWMutex
+	lock     sync.RWMutex
+	lockHdlr sync.RWMutex
 
 	role ICERole
 	// Component ICEComponent
@@ -167,15 +168,15 @@ func (t *ICETransport) Stop() error {
 // OnSelectedCandidatePairChange sets a handler that is invoked when a new
 // ICE candidate pair is selected
 func (t *ICETransport) OnSelectedCandidatePairChange(f func(*ICECandidatePair)) {
-	t.lock.Lock()
-	defer t.lock.Unlock()
+	t.lockHdlr.Lock()
+	defer t.lockHdlr.Unlock()
 	t.onSelectedCandidatePairChangeHdlr = f
 }
 
 func (t *ICETransport) onSelectedCandidatePairChange(pair *ICECandidatePair) {
-	t.lock.RLock()
+	t.lockHdlr.RLock()
 	hdlr := t.onSelectedCandidatePairChangeHdlr
-	t.lock.RUnlock()
+	t.lockHdlr.RUnlock()
 	if hdlr != nil {
 		hdlr(pair)
 	}
@@ -184,15 +185,15 @@ func (t *ICETransport) onSelectedCandidatePairChange(pair *ICECandidatePair) {
 // OnConnectionStateChange sets a handler that is fired when the ICE
 // connection state changes.
 func (t *ICETransport) OnConnectionStateChange(f func(ICETransportState)) {
-	t.lock.Lock()
-	defer t.lock.Unlock()
+	t.lockHdlr.Lock()
+	defer t.lockHdlr.Unlock()
 	t.onConnectionStateChangeHdlr = f
 }
 
 func (t *ICETransport) onConnectionStateChange(state ICETransportState) {
-	t.lock.RLock()
+	t.lockHdlr.RLock()
 	hdlr := t.onConnectionStateChangeHdlr
-	t.lock.RUnlock()
+	t.lockHdlr.RUnlock()
 	if hdlr != nil {
 		hdlr(state)
 	}

--- a/icetransport.go
+++ b/icetransport.go
@@ -6,6 +6,7 @@ import (
 	"context"
 	"errors"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/pion/ice"
@@ -16,16 +17,15 @@ import (
 // ICETransport allows an application access to information about the ICE
 // transport over which packets are sent and received.
 type ICETransport struct {
-	lock     sync.RWMutex
-	lockHdlr sync.RWMutex
+	lock sync.RWMutex
 
 	role ICERole
 	// Component ICEComponent
 	// State ICETransportState
 	// gatheringState ICEGathererState
 
-	onConnectionStateChangeHdlr       func(ICETransportState)
-	onSelectedCandidatePairChangeHdlr func(*ICECandidatePair)
+	onConnectionStateChangeHdlr       atomic.Value // func(ICETransportState)
+	onSelectedCandidatePairChangeHdlr atomic.Value // func(*ICECandidatePair)
 
 	state ICETransportState
 
@@ -168,34 +168,26 @@ func (t *ICETransport) Stop() error {
 // OnSelectedCandidatePairChange sets a handler that is invoked when a new
 // ICE candidate pair is selected
 func (t *ICETransport) OnSelectedCandidatePairChange(f func(*ICECandidatePair)) {
-	t.lockHdlr.Lock()
-	defer t.lockHdlr.Unlock()
-	t.onSelectedCandidatePairChangeHdlr = f
+	t.onSelectedCandidatePairChangeHdlr.Store(f)
 }
 
 func (t *ICETransport) onSelectedCandidatePairChange(pair *ICECandidatePair) {
-	t.lockHdlr.RLock()
-	hdlr := t.onSelectedCandidatePairChangeHdlr
-	t.lockHdlr.RUnlock()
+	hdlr := t.onSelectedCandidatePairChangeHdlr.Load()
 	if hdlr != nil {
-		hdlr(pair)
+		hdlr.(func(*ICECandidatePair))(pair)
 	}
 }
 
 // OnConnectionStateChange sets a handler that is fired when the ICE
 // connection state changes.
 func (t *ICETransport) OnConnectionStateChange(f func(ICETransportState)) {
-	t.lockHdlr.Lock()
-	defer t.lockHdlr.Unlock()
-	t.onConnectionStateChangeHdlr = f
+	t.onConnectionStateChangeHdlr.Store(f)
 }
 
 func (t *ICETransport) onConnectionStateChange(state ICETransportState) {
-	t.lockHdlr.RLock()
-	hdlr := t.onConnectionStateChangeHdlr
-	t.lockHdlr.RUnlock()
+	hdlr := t.onConnectionStateChangeHdlr.Load()
 	if hdlr != nil {
-		hdlr(state)
+		hdlr.(func(ICETransportState))(state)
 	}
 }
 

--- a/peerconnection_close_test.go
+++ b/peerconnection_close_test.go
@@ -102,3 +102,65 @@ func TestPeerConnection_Close_PreICE(t *testing.T) {
 		time.Sleep(time.Second)
 	}
 }
+
+func TestPeerConnection_Close_DuringICE(t *testing.T) {
+	// Limit runtime in case of deadlocks
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
+
+	report := test.CheckRoutines(t)
+	defer report()
+
+	pcOffer, pcAnswer, err := newPair()
+	if err != nil {
+		t.Fatal(err)
+	}
+	closedOffer := make(chan struct{})
+	closedAnswer := make(chan struct{})
+	pcAnswer.OnICEConnectionStateChange(func(iceState ICEConnectionState) {
+		if iceState == ICEConnectionStateConnected {
+			go func() {
+				if err2 := pcAnswer.Close(); err2 != nil {
+					t.Errorf("pcAnswer.Close() failed: %v", err2)
+				}
+				close(closedAnswer)
+				if err2 := pcOffer.Close(); err2 != nil {
+					t.Errorf("pcOffer.Close() failed: %v", err2)
+				}
+				close(closedOffer)
+			}()
+		}
+	})
+
+	offer, err := pcOffer.CreateOffer(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = pcOffer.SetLocalDescription(offer); err != nil {
+		t.Fatal(err)
+	}
+	if err = pcAnswer.SetRemoteDescription(offer); err != nil {
+		t.Fatal(err)
+	}
+	answer, err := pcAnswer.CreateAnswer(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = pcAnswer.SetLocalDescription(answer); err != nil {
+		t.Fatal(err)
+	}
+	if err = pcOffer.SetRemoteDescription(answer); err != nil {
+		t.Fatal(err)
+	}
+
+	select {
+	case <-closedAnswer:
+	case <-time.After(5 * time.Second):
+		t.Error("pcAnswer.Close() Timeout")
+	}
+	select {
+	case <-closedOffer:
+	case <-time.After(5 * time.Second):
+		t.Error("pcOffer.Close() Timeout")
+	}
+}

--- a/peerconnection_media_test.go
+++ b/peerconnection_media_test.go
@@ -210,7 +210,8 @@ This test adds an input track and asserts
 * No deadlocks on shutdown
 */
 func TestPeerConnection_Media_Shutdown(t *testing.T) {
-	iceComplete := make(chan bool)
+	iceCompleteAnswer := make(chan struct{})
+	iceCompleteOffer := make(chan struct{})
 
 	lim := test.TimeOut(time.Second * 30)
 	defer lim.Stop()
@@ -250,7 +251,7 @@ func TestPeerConnection_Media_Shutdown(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	var onTrackFiredLock sync.RWMutex
+	var onTrackFiredLock sync.Mutex
 	onTrackFired := false
 
 	pcAnswer.OnTrack(func(track *Track, receiver *RTPReceiver) {
@@ -261,9 +262,12 @@ func TestPeerConnection_Media_Shutdown(t *testing.T) {
 
 	pcAnswer.OnICEConnectionStateChange(func(iceState ICEConnectionState) {
 		if iceState == ICEConnectionStateConnected {
-			go func() {
-				close(iceComplete)
-			}()
+			close(iceCompleteAnswer)
+		}
+	})
+	pcOffer.OnICEConnectionStateChange(func(iceState ICEConnectionState) {
+		if iceState == ICEConnectionStateConnected {
+			close(iceCompleteOffer)
 		}
 	})
 
@@ -271,7 +275,8 @@ func TestPeerConnection_Media_Shutdown(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	<-iceComplete
+	<-iceCompleteAnswer
+	<-iceCompleteOffer
 
 	// Each PeerConnection should have one sender, one receiver and two transceivers
 	for _, pc := range []*PeerConnection{pcOffer, pcAnswer} {

